### ISSUE TITLE
Add boost to official repo

### ISSUE THREAD

**Issue by Feynstein** [mrc-ide/covid-sim#229]
It's been talked in another issue (#202) that in order to make any significant improvement to the code boost should be integrated to this library. It's used in most scientific code I know about. It contains everything that already made it to  C++17 and more. It would be a better way to get good functionality rather than upgrading to C++17.

**Comment by zebmason** [comment]
In #220 I mentioned the possibility of using [Boost Geometry](https://www.boost.org/doc/libs/1_73_0/libs/geometry/doc/html/index.html) even though I rather dislike Boost. Apart from filesystem, for which there is a backport of std::filesystem, what significant improvements are you proposing that Boost would facilitate?

**Comment by Feynstein** [comment]
Well geometry is one. From the top of my head there is program options, splines for interpolation and the fact they can lower the size of data when lossless, and linear algebra... And I don't know how they handle their 3D geography but maybe something that can handle meshes.  I'll check other code I do and see what could apply.

**Comment by Feynstein** [comment]
I think that maybe OCTrees could be used for performance... But I don't know if it's handled in boost. I'm not on my PC right now.

**Comment by Feynstein** [comment]
@zebmason heres a short list of my usual includes in cmake
atomic
		chrono
		context
		coroutine
		date_time
		filesystem
		graph
		iostreams
		locale
		log_setup
		log
		math_c99f
		math_c99l
		math_c99
		math_tr1f
		math_tr1l
		math_tr1
		program_options
		random
		serialization
		#signals    #no longer exists in boost 1.72
		system
		thread
		timer
		#type_erasure
		wserialization)

**Comment by Feynstein** [comment]
Some of which I don't personally use but that my computer science colleagues use

**Comment by Feynstein** [comment]
Eigen, PCL or VTK (Paraview for visualization) and even opencv (you can do other stuff than only imaging with it) could be good additions also... And yes, there are other libraries being used but they are not named explicitely in this list because all they are included somewhere esle or something.

**Comment by matt-gretton-dann** [comment]
To repeat what I said in #220 the requirements I have on any third-part library are:

* Works on Windows & UNIX like (Linux-based & macOS at least) platforms.
* License compatible with GPLv3 when statically linked
* No install steps required on any platform over and above copying a single CovidSim executable.
* Simple to integrate to the current build systems (in particular Visual Studio) - by which I mean no extra steps after cloning the repo if at all possible.
* Can be implemented without major changes to the current usage models.

These are to ensure that development of model features is not slowed down by re-engineering going on at the same time. I acknowledge that the above may be non-optimal, but the model is in active development and use, and so I'm being conservative about significant changes.

I would also like to keep the number of libraries down to a sensible minimum, as it makes managing the repo simpler.

I do not think it would be appropriate to add libraries based on the off-chance of using them later.  If using a particular library will make life easier then add it in the PR which makes the other change.

Which isn't to say I object to Boost - I just need to see a use case where it (or one of the component libraries) bring a tangible benefit.

**Comment by Feynstein** [comment]
@matt-gretton-dann  ... also... I just thought that maybe you could start an official experimental branch where these kind of things could be approved without damaging the main model. It could be run in parallel to the real model somewhere else to check its outputs.

**Comment by weshinsley** [comment]
Cleaning up on this one: conclusion is that we'll add dependencies only if there's a very well-defined use-case (ie, add the dependency along with the PR that uses it). argv handling was implemented in PR #228 and the experimental fork discussion was decided in issue #231.
